### PR TITLE
Support generating with fallback for short form audio in Whisper

### PR DESCRIPTION
**Pull request by kamilakesbi**
# What does this PR do?

The aim of this PR is to refacto the Whisper `generate` method to handle both short form and long form audio generation similarly. It will support short form audio generation with fallback (as requested in #29508). 

Here's what I've done: 

### Removed previous short-form scripts: 

I've removed the part of the code used for short form generation. This involve lines [562 to 603 ](https://github.com/huggingface/transformers/blob/348e2294ac1a3cc744a93f04ab1f886439ee90d3/src/transformers/models/whisper/generation_whisper.py#L562) and lines [498 to 505](https://github.com/huggingface/transformers/blob/348e2294ac1a3cc744a93f04ab1f886439ee90d3/src/transformers/models/whisper/generation_whisper.py#L498) in main. Now when a short form audio (or a batched short form of audio) is passed to `generate`, it's processed by the part of the code previously used for long form generation.

### Use is_shortform to still distinguish between short form and long form in some cases: 

- In the `_postprocess_outputs` method we only return `past_key_values` if the audios are short form. For long form audios it is too expensive. (cf. this [line](https://github.com/huggingface/transformers/blob/80c5bc6a212a39e8a3ccb340ef5a7a4c2609f5e6/src/transformers/models/whisper/generation_whisper.py#L932)). 

-  In `_retrieve_max_frames_and_seek `:  For long form audios, we necessarily need to pass an attention mask but not for short form audios. We can thus compute `max_frames` and `seek` without relying on the attention mask for short form audios.

- I've also updated the `split_by_batch_index` method:  the previous method was broken when return_dict_in_generate was set to True for different short form audio cases. Now it handles both short form and long form audios. 

- I've removed the `is_shortform` parameter from the inputs to the `_retrieve_logit_processors ` method to allow the use of `generation_config.no_speech_threshold` for short form audios.  

- I've removed `is_shortfrom` parameter from the inputs to the `_set_return_outputs ` method to allow the use of `logprob_threshold` for short form audios. 

### Make num_return_sequences>1 compatible with generate_with_fallback: 

- This is a bit tricky because `generate_with_fallback` can't handle num_return_sequences>1 by design. I've added a new method, called `_expand_variables_for_generation` , which expands the different variables before passing into `generate_with_fallback` when `generation_config.num_return_sequences>1`. After expansion it will set `generation_config.num_return_sequences` to 1 for compatibility with `generate_with_fallback`. 

### Ensure that the output format for short form audio is compatible with the output format in main: 

The output format for long-form audio is different from that for short-form audio. In order to ensure that the output is similar to that obtained in main when processing short form audio, we need to add a few post-processing steps: This is what is done in lines [721 to 765](https://github.com/huggingface/transformers/blob/80c5bc6a212a39e8a3ccb340ef5a7a4c2609f5e6/src/transformers/models/whisper/generation_whisper.py#L721). In particular here: 

- We add an `EOS` token to the output sequence as it was removed during generation with fallback. 
-  We return the token timestamps if `return_token_timestamps` is True in the correct format ([see here](https://github.com/huggingface/transformers/blob/80c5bc6a212a39e8a3ccb340ef5a7a4c2609f5e6/src/transformers/models/whisper/generation_whisper.py#L741)). 
- If `return_dict_in_generate` is True, we use the new method `_stack_split_outputs` to get the output dict (containing all attributes (scores, encoder_attentions, etc.)) in the right format. `_stack_split_outputs` basically performs the opposite operations to `split_by_batch_index `.

### Make failing slow tests to pass: 

- I've updated some failing slow tests and made them pass ([see here](https://github.com/huggingface/transformers/pull/30984#issuecomment-2148028809)). 

### Add new tests to make sure generation with fallback works for short form audios: 

I've added two tests:  `test_whisper_shortform_single_batch_prev_cond ` and `test_whisper_shortform_multi_batch_hard_prev_cond`. 

## Who can review: 

@sanchit-gandhi

### CODE REVIEW

File: src/transformers/models/whisper/generation_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:667]: Why only with long-form?

File: src/transformers/models/whisper/generation_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:711]: Surely fine to do with short-form as well?

File: src/transformers/models/whisper/generation_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:797]: Should we handle the case where `is_shortform` and `return_segments=True`? Here, I would return the same output as long-form with `return_segments=True`

**Review by sanchit-gandhi — COMMENTED** [review]
Looks like a good start @kamilakesbi. Two biggest suggestions are related to the designs of i) assisted generation, and ii) num return sequences. Think both can be simplified and assisted generation made more rigorous.

Two further design questions:
1. Should we return the original `decoder_input_ids` and EOS tokens in the sequences for long-form generation as well? IMO this is an inconsistency that we return them for short-form, but not long-form, and I would be in-favour of unifying the two in this PR
2. Is it correct to de-activate beam search when `temperature>0`? We currently don't do this for long-form generation, but given the original Whisper repo does, it would be good to determine whether this is a 'bug' or an intended design decision

File: src/transformers/models/whisper/generation_whisper.py
Comment (kamilakesbi) [reply src/transformers/models/whisper/generation_whisper.py:612]: I see! in practice several tests failed because `generation_config.no_timestamps_token_id` isn't defined for some models. With the prior short form generation these tests passed because `no_timestamps_token_id` wasn't used. 
It could be related to issue #21878. 

Instead of `if not is_shortform:` we could perhaps use: `if generation_config.no_timestamps_token_id is not None:` for now...

File: src/transformers/models/whisper/generation_whisper.py
Comment (kamilakesbi) [reply src/transformers/models/whisper/generation_whisper.py:1509]: Agree, only we should do instead:  

```diff
-max_frames = torch.ones((1,), dtype=torch.long) * total_input_frames
-seek = torch.zeros((1,), dtype=torch.long)
+max_frames = torch.ones((batch_size, ), dtype=torch.long) * total_input_frames
+seek = torch.zeros((batch_size,), dtype=torch.long)
```

**Comment by kamilakesbi** [comment]
I have run `test_modeling_whisper` slow tests. Here are the currently failing tests (which pass on main): 

- test_assisted_decoding_matches_greedy_search_0_random
- test_speculative_decoding_non_distil
- test_generate_continue_from_past_key_values
- test_tiny_en_generation
- test_tiny_generation
- test_tiny_en_batched_generation
- test_tiny_timestamp_generation 
- test_tiny_token_timestamp_batch_generation
- test_large_timestamp_generation

File: src/transformers/models/whisper/generation_whisper.py
Comment (sanchit-gandhi) [reply src/transformers/models/whisper/generation_whisper.py:886]: Not sure why this was marked as resolved, but not implemented in the latest commit:
https://github.com/huggingface/transformers/blob/38aa42d459ad5a904077ecf683fb5b670f4f2540/src/transformers/models/whisper/generation_whisper.py#L886

For all comments, you should either:
1. Implement the requested changes (if you agree with the proposition); or
2. Reply with a comment explaining why the changes should not be implemented or if an alternative design is required

=> only then should a comment be marked as resolved

Otherwise, when doing a second review, I'm just going to make the same comments as my first review, since I don't know why you haven't accepted them. This is a waste of time for both of us!

File: src/transformers/models/whisper/generation_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:554]: Do you think it would be cleaner to put this logic into `_set_return_timestamps` and return the timestamp token? This would keep the main body of the generate function cleaner

```python
timestamp_begin = _set_return_timestamps(...)
```

File: src/transformers/models/whisper/generation_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:1623]: Changes should be able to be reverted after setting `timestamp_begin = self.config.vocab_size + 1`, since then no element in `seek_sequence` can ever be `ge` than `timestamp_begin`

File: src/transformers/models/whisper/generation_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:610]: Is this `if/else` logic required? It should be handled by `_get_attr_from_logig_processors` right?
https://github.com/huggingface/transformers/blob/38aa42d459ad5a904077ecf683fb5b670f4f2540/src/transformers/models/whisper/generation_whisper.py#L121-L123

File: src/transformers/models/whisper/generation_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:968]: Can you refer to `_expand_inputs_for_generation` to see how this can potentially be done in a cleaner way (especially for the torch tensors)?

https://github.com/huggingface/transformers/blob/38aa42d459ad5a904077ecf683fb5b670f4f2540/src/transformers/generation/utils.py#L597C9-L597C38

File: src/transformers/models/whisper/generation_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:618]: So we pop them here and then don't do anything with them? Is this intended? Shouldn't we pop them from the kwargs, and if they're None, then initialise them from the generation config? 

This logic should go in `_prepare_decoder_input_ids` either way

File: src/transformers/models/whisper/generation_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:634]: Logic can go in `_set_max_new_tokens_and_length`

File: src/transformers/models/whisper/generation_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:871]: Would be cleaner **not** to return `seek_outputs_short_form` here, since we make a copy of `seek_outputs`, and then just do any necessary post-processing in `_postprocess_outputs` or the main body of generate

File: src/transformers/models/whisper/generation_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:659]: Strongly think there shouldn't be a need for `seek_outputs_short_form` here - should be possible to handle this in `_postprocess_outputs` and the main generate body here

File: src/transformers/models/whisper/generation_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:717]: Don't know why we still do this rather than not stripping the input ids in the first place https://github.com/huggingface/transformers/pull/30984#discussion_r1618595148

File: src/transformers/models/whisper/generation_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:731]: Why do we do this?

File: tests/models/whisper/test_modeling_whisper.py
Comment (sanchit-gandhi) [inline tests/models/whisper/test_modeling_whisper.py:1542]: Can you add a comment above the test definition explaining why we need to override this test?

**Review by sanchit-gandhi — COMMENTED** [review]
Design looks much better, thanks for iterating @kamilakesbi! Mostly a clean-up job now

**Comment by kamilakesbi** [comment]
@sanchit-gandhi I have taken all your reviews into account and made sure that the CI tests and slow tests have passed. 

This PR should be ready for final review!

cc @amyeroberts.

**Comment by kamilakesbi** [comment]
Added two slow tests (`test_whisper_shortform_single_batch_prev_cond` and `test_whisper_shortform_multi_batch_hard_prev_cond`) and made a few modifications to make them pass.

File: src/transformers/models/whisper/generation_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:912]: Could we first set a `start_idx` as follows:
```python
start_idx = decoder_input_ids.shape[-1] if not is_shortform else torch.tensor(0)
```

And then remove all the `if/else` statements on `is_shortform` in the rest of the method?

E.g. the following:
```python
            if not is_shortform:
                seek_outputs = seek_outputs[:, decoder_input_ids.shape[-1] :]
                return seek_outputs, seek_outputs
            else:
                return seek_outputs, seek_outputs
```

Becomes:
```python
            seek_outputs = seek_outputs[:, start_idx :]
            return seek_outputs, seek_outputs
```

Same for the token timestamps and sequences later in the method

File: src/transformers/models/whisper/generation_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:723]: Probably possible to keep the EOS token directly in the generated sequence as suggested in the previous comment

File: src/transformers/models/whisper/generation_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:953]: Only nit with this method is that we stack everything on CPU, and then move over to GPU again

=> could you explore quickly whether there's a way to create the stacked tensors directly on device?

Otherwise, the CPU <-> GPU transfer is going to be quite slow

File: src/transformers/models/whisper/generation_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:1710]:
```suggestion
```

File: tests/models/whisper/test_modeling_whisper.py
Comment (sanchit-gandhi) [inline tests/models/whisper/test_modeling_whisper.py:2159]: Why do we get these extra tokens here?

**Review by sanchit-gandhi — APPROVED** [review]
Think we can clean-up `_stack_split_outputs` a bit and avoid cropping then re-inserting the EOS token ids, otherwise looks good to me! After these changes, feel free to request a review directly from Arthur - it'd be great to get a second pair of eyes from someone who's worked a lot on the Whisper codebase as well.

**Comment by kamilakesbi** [comment]
Hi @ArthurZucker, could you please do a review on this PR ? 
Failing tests on the CI are unrelated to this work I think :)

**Comment by kamilakesbi** [comment]
Hi @amy, in this PR we only focus on the unification of the long and short form audio logic! we don't change how speculative decoding is handled here :)

**Comment by amyeroberts** [comment]
> I've adapted the code to make it compatible with speculative decoding, return_tokens_timestamps, return_timestamps, return_dict_in_generate, num_beams>=1 and num_return_sequences>=1, max_new_tokens and max_length parameters.

@kamilakesbi Could you update the PR description to reflect the current state (either disambiguate for confused people like me or remove :) )?

**Comment by ArthurZucker** [comment]
I'll take care of reviewing it in a bit!

**Comment by kamilakesbi** [comment]
Gentle ping @ArthurZucker

File: src/transformers/models/whisper/generation_whisper.py
Comment (ArthurZucker) [inline src/transformers/models/whisper/generation_whisper.py:576]: `num_return_sequences  = generation_config.num_return_sequences(..........) = self._expand_variables_for_generation(...)` find the `generation_config.num_return_sequences(..........) ` a bit weird but alright

File: src/transformers/models/whisper/generation_whisper.py
Comment (ArthurZucker) [inline src/transformers/models/whisper/generation_whisper.py:736]: let's split this in 2 lines

File: src/transformers/models/whisper/generation_whisper.py
Comment (ArthurZucker) [reply src/transformers/models/whisper/generation_whisper.py:843]: is this adressed?

File: src/transformers/models/whisper/generation_whisper.py
Comment (ArthurZucker) [reply src/transformers/models/whisper/generation_whisper.py:949]: is this resolved?

Comment (ArthurZucker) [reply src/transformers/models/whisper/generation_whisper.py:949]: also would be nice to documment 1 ligne about why we need this method

File: src/transformers/models/whisper/generation_whisper.py
Comment (ArthurZucker) [inline src/transformers/models/whisper/generation_whisper.py:1030]: looks duplicate

File: src/transformers/models/whisper/generation_whisper.py
Comment (ArthurZucker) [inline src/transformers/models/whisper/generation_whisper.py:1038]: here as well

File: tests/models/whisper/test_modeling_whisper.py
Comment (ArthurZucker) [inline tests/models/whisper/test_modeling_whisper.py:1979]: why are these updated?

File: tests/models/whisper/test_modeling_whisper.py
Comment (ArthurZucker) [inline tests/models/whisper/test_modeling_whisper.py:2806]: would be nice to have both with and without here to see what this is acutally bringing to the table, because for all I know the expected text here is the same as without

**Review by ArthurZucker — COMMENTED** [review]
Great work! Sorry for my late review, this seems to add a lot of extra logic, but necessary so let's ship this! 🔥 

Overall LGTM, let's make sure we don't break BC and we can merge (test being modified 😨

**Comment by kamilakesbi** [comment]
@ArthurZucker thanks for your review! I took your remarks into account :) 

Failing tests are unrelated to this PR. If this is ok for you we can perhaps merge or wait for the CI to be green...

**Comment by ArthurZucker** [comment]
Let's wait for the full CI seems alright now!

File: tests/models/whisper/test_modeling_whisper.py
Comment (ArthurZucker) [reply tests/models/whisper/test_modeling_whisper.py:2020]: THis is not answered

**Comment by ArthurZucker** [comment]
Also a question ont answered!

**Comment by kamilakesbi** [comment]
The CI is green yes :) if it's ok for you I can merge!

File: src/transformers/models/whisper/generation_whisper.py
Comment (ArthurZucker) [inline src/transformers/models/whisper/generation_whisper.py:732]: here you should decalre `somthing =torch.full((sequences.shape[0], 1,), generation_config.eos_token_id).to(sequences.device),` then do sequences = torch.cat

File: tests/models/whisper/test_modeling_whisper.py
Comment (ArthurZucker) [reply tests/models/whisper/test_modeling_whisper.py:2020]: This is still not answered

File: src/transformers/models/whisper/generation_whisper.py
Comment (ArthurZucker) [inline src/transformers/models/whisper/generation_whisper.py:732]:
```suggestion
                eos_tokens = torch.full((sequences.shape[0], 1), generation_config.eos_token_id)
                sequences = torch.cat([sequences, eos_tokens.to(sequences.device)], dim=-1)
```

File: src/transformers/models/whisper/generation_whisper.py
Comment (ArthurZucker) [inline src/transformers/models/whisper/generation_whisper.py:754]: this one is not protected by the num_return_sequences

**Review by ArthurZucker — APPROVED** [review]
Thanks! Last to nits and you can merge!

**Pull request merged**
